lib/puppet/pops/lookup/lookup_config.rb

```diff
@@ -0,0 +1,211 @@
+        @config[KEY_HIERARCHY] ||= DEFAULT_CONFIG[KEY_HIERARCHY]
+        @config[KEY_DATADIR] ||= DEFAULT_CONFIG[KEY_DATADIR]
+      else
+        @config = DEFAULT_CONFIG
+        @config_path = nil
+      end
+    end
+    @version = @config[KEY_VERSION]
+  end
+
+  # Creates the data providers for this config
+  #
+  # @param lookup_invocation [Invocation] Invocation data containing scope, overrides, and defaults
+  # @param parent_data_provider [DataProvider] The data provider that loaded this configuration
+  # @return [Array<DataProvider>] the created providers
+  def create_configured_data_providers(lookup_invocation, parent_data_provider)
+    datadir = @config.include?('datadir') ? @config_root + @config['datadir'] : nil
+
+    # Hashes enumerate their values in the order that the corresponding keys were inserted so it's safe to use
+    # a hash for the data_providers.
+    data_providers = {}
+    @config[KEY_HIERARCHY].each do |he|
+      name = he[KEY_NAME]
+      raise Puppet::DataBinding::LookupError, "#{@config_path}: Name '#{name}' defined more than once" if data_providers.include?(name)
+      function_kind = FUNCTION_KEYS.find { |key| he.include?(key) }
+      function_name = he[function_kind]
+
+      location_key = LOCATION_KEYS.find { |key| he.include?(key) }
+      locations = case location_key
+      when KEY_PATHS
+        resolve_paths(datadir, he[location_key], lookup_invocation)
+      when KEY_PATH
+        resolve_paths(datadir, [he[location_key]], lookup_invocation)
+      when KEY_GLOBS
+        expand_globs(datadir, he[location_key], lookup_invocation)
+      when KEY_GLOB
+        expand_globs(datadir, [he[location_key]], lookup_invocation)
+      when KEY_URIS
+        expand_uris(he[location_key], lookup_invocation)
+      when KEY_URI
+        expand_uris([he[location_key]], lookup_invocation)
+      else
+        nil
+      end
+      options = he[KEY_OPTIONS]
+      options = options.nil? ? EMPTY_HASH : interpolate(options, lookup_invocation, false)
+      data_providers[name] = create_data_provider(name, parent_data_provider, function_kind, function_name, options, locations)
+    end
+    data_providers.values
+  end
+
+  def name
+    "lookup version #{version}"
+  end
+
+  private
+
+  def create_data_provider(name, parent_data_provider, function_kind, function_name, options, locations)
+    FUNCTION_PROVIDERS[function_kind].new(name, parent_data_provider, function_name, options, locations)
+  end
+
+  def validate_config(config)
+    Types::TypeAsserter.assert_instance_of(["The Lookup Configuration at '%s'", @config_path], self.class.config_type, config)
+    config[KEY_HIERARCHY].each do |he|
+      name = he[KEY_NAME]
+      case FUNCTION_KEYS.count { |key| he.include?(key) }
+      when 1
+        # OK
+      when 0
+        raise Puppet::DataBinding::LookupError,
+          "#{@config_path}: One of #{combine_strings(FUNCTION_KEYS - [KEY_LEGACY_DATA_HASH])} must defined in hierarchy '#{name}'"
```

Comment:
The error message is prefixed with the full path of the file which makes it easy to determine if it concerns an environment or a module. The file in question then has an explicit `hierarchy:` key consisting of named entries. Referring to such entries as "in hierarchy 'common'" should be very clear IMHO. Clearer than "lookup configuration entry 'common'".